File: aac/api/python/wia_aac/core/wia_aac.py
"""
WIA AAC Main Class
Primary API entry point for AAC sensor interaction
"""

import time
import logging
from typing import Optional, List, Callable, Any, Dict, Type

from .event_emitter import EventEmitter
from ..types import (
    WiaAacSignal,
    DeviceInfo,
    SensorType,
    EventType,
    ErrorCode,
    WiaAacError,
    SelectionEvent,
    DisconnectReason,
    SensorConfig,
    WiaAacOptions,
    ConnectionState,
    SensorOptions
)


# Adapter registry
_adapter_registry: Dict[str, Type["BaseSensorAdapter"]] = {}


class WiaAac:
    """WIA AAC API main class"""

    def __init__(self, options: Optional[WiaAacOptions] = None):
        self._options = options or WiaAacOptions()
        self._emitter = EventEmitter()
        self._adapter: Optional["BaseSensorAdapter"] = None
        self._config: Optional[SensorConfig] = None
        self._state = ConnectionState.DISCONNECTED
        self._signal_buffer: List[WiaAacSignal] = []
        self._logger = self._create_logger()

    # ============================================
    # Static Methods
    # ============================================

    @staticmethod
    def register_adapter(sensor_type: str, adapter_class: Type["BaseSensorAdapter"]) -> None:
        """Register a custom adapter"""
        _adapter_registry[sensor_type] = adapter_class

    @staticmethod
    def get_adapter_class(sensor_type: str) -> Optional[Type["BaseSensorAdapter"]]:
        """Get registered adapter class"""
        return _adapter_registry.get(sensor_type)

    # ============================================
    # Connection Management
    # ============================================

    async def connect(self, config: SensorConfig) -> None:
        """Connect to a sensor"""
        if self._state == ConnectionState.CONNECTED:
            raise WiaAacError(
                code=ErrorCode.DEVICE_BUSY,
                message="Already connected to a device",
                timestamp=int(time.time() * 1000),
                recoverable=True
            )

        self._config = config
        self._state = ConnectionState.CONNECTING
        sensor_type = config.type.value if isinstance(config.type, SensorType) else config.type
        self._logger.info(f"Connecting to {sensor_type} sensor...")

        try:
            # Get adapter (use pre-set adapter if available)
            if not self._adapter:
                adapter_class = _adapter_registry.get(sensor_type)
                if not adapter_class:
                    self._logger.warning(f"No adapter registered for {sensor_type}, using mock")
                    from ..adapters.mock_adapter import MockAdapter
                    self._adapter = MockAdapter(sensor_type=sensor_type)
                else:
                    self._adapter = adapter_class()

            # Connect
            await self._adapter.connect(config)

            # Setup signal handler
            self._adapter.on_signal(self._handle_signal)

            # Apply options
            if config.options:
                self._adapter.configure(config.options)

            self._state = ConnectionState.CONNECTED
            manufacturer = config.device.manufacturer if config.device else sensor_type
            self._logger.info(f"Connected to {manufacturer}")

            # Emit connected event
            device_info = self._adapter.device_info or DeviceInfo(
                manufacturer=config.device.manufacturer if config.device else "Unknown",
                model=config.device.model if config.device else sensor_type
            )
            self._emitter.emit(EventType.CONNECTED, device_info)

        except Exception as e:
            self._state = ConnectionState.ERROR
            error = WiaAacError(
                code=ErrorCode.CONNECTION_FAILED,
                message=f"Failed to connect: {str(e)}",
                timestamp=int(time.time() * 1000),
                recoverable=True
            )
            self._emitter.emit(EventType.ERROR, error)
            raise error

    async def disconnect(self) -> None:
        """Disconnect from the sensor"""
        if self._adapter:
            try:
                await self._adapter.disconnect()
            except Exception as e:
                self._logger.error(f"Error during disconnect: {e}")
            self._adapter = None

        self._state = ConnectionState.DISCONNECTED
        self._config = None
        self._signal_buffer = []

        self._emitter.emit(EventType.DISCONNECTED, DisconnectReason(reason="user"))
        self._logger.info("Disconnected")

    def is_connected(self) -> bool:
        """Check if connected"""
        return self._state == ConnectionState.CONNECTED

    def get_connection_state(self) -> ConnectionState:
        """Get connection state"""
        return self._state

    # ============================================
    # Device Discovery
    # ============================================

    async def list_devices(self) -> List[DeviceInfo]:
        """List available devices"""
        self._logger.debug("Listing devices...")
        return []

    def get_device_info(self) -> Optional[DeviceInfo]:
        """Get connected device info"""
        return self._adapter.device_info if self._adapter else None

    # ============================================
    # Event Handling
    # ============================================

    def on(self, event: str, handler: Callable[[Any], None]) -> None:
        """Subscribe to an event"""
        self._emitter.on(event, handler)

    def off(self, event: str, handler: Callable[[Any], None]) -> None:
        """Unsubscribe from an event"""
        self._emitter.off(event, handler)

    def once(self, event: str, handler: Callable[[Any], None]) -> None:
        """Subscribe to an event once"""
        self._emitter.once(event, handler)

    # ============================================
    # Signal Management
    # ============================================

    def get_last_signal(self) -> Optional[WiaAacSignal]:
        """Get the last received signal"""
        return self._signal_buffer[-1] if self._signal_buffer else None

    def get_signal_buffer(self) -> List[WiaAacSignal]:
        """Get signal buffer"""
        return self._signal_buffer.copy()

    # ============================================
    # Configuration
    # ============================================

    def configure(self, options: SensorOptions) -> None:
        """Update sensor options"""
        if self._adapter:
            self._adapter.configure(options)

    def get_config(self) -> Optional[SensorConfig]:
        """Get current configuration"""
        return self._config

    # ============================================
    # Private Methods
    # ============================================

    def _handle_signal(self, signal: WiaAacSignal) -> None:
        """Handle incoming signal"""
        # Add to buffer
        self._signal_buffer.append(signal)
        if len(self._signal_buffer) > self._options.signal_buffer_size:
            self._signal_buffer.pop(0)

        # Emit signal event
        self._emitter.emit(EventType.SIGNAL, signal)

        # Check for selection events
        self._check_selection_event(signal)

    def _check_selection_event(self, signal: WiaAacSignal) -> None:
        """Check if signal triggers a selection event"""
        dwell_time = 1000
        if self._config and self._config.options:
            if hasattr(self._config.options, 'dwell_time') and self._config.options.dwell_time:
                dwell_time = self._config.options.dwell_time

        # Eye tracker fixation -> selection
        if isinstance(signal.data, dict):
            data = signal.data
        else:
            return

        signal_type = signal.type.value if isinstance(signal.type, SensorType) else signal.type

        if signal_type == "eye_tracker":
            fixation = data.get("fixation", {})
            if fixation.get("active") and fixation.get("duration_ms", 0) >= dwell_time:
                gaze = data.get("gaze", {})
                self._emitter.emit(EventType.SELECTION, SelectionEvent(
                    timestamp=signal.timestamp.unix_ms,
                    target_id=fixation.get("target_id", "unknown"),
                    target_type="area",
                    selection_method="dwell",
                    position={"x": gaze.get("x", 0), "y": gaze.get("y", 0)},
                    confidence=signal.meta.confidence if signal.meta else 0.8
                ))

        # Switch press -> selection
        elif signal_type == "switch" and data.get("state") == "pressed":
            self._emitter.emit(EventType.SELECTION, SelectionEvent(
                timestamp=signal.timestamp.unix_ms,
                target_id=f"switch_{data.get('switch_id', 1)}",
                target_type="button",
                selection_method="switch",
                confidence=1.0
            ))

    def _create_logger(self) -> logging.Logger:
        """Create logger"""
        logger = logging.getLogger("WiaAac")

        level_map = {
            "debug": logging.DEBUG,
            "info": logging.INFO,
            "warn": logging.WARNING,
            "error": logging.ERROR,
            "none": logging.CRITICAL + 1
        }

        logger.setLevel(level_map.get(self._options.log_level, logging.INFO))

        if not logger.handlers:
            handler = logging.StreamHandler()
            handler.setFormatter(logging.Formatter("[WiaAac:%(levelname)s] %(message)s"))
            logger.addHandler(handler)

        return logger

    def use_adapter(self, adapter: "BaseSensorAdapter") -> None:
        """Use a specific adapter instance (for testing)"""
        self._adapter = adapter


# Import at the end to avoid circular imports
from ..adapters.base_adapter import BaseSensorAdapter

File: aac/api/typescript/src/core/WiaAac.ts
/**
 * WIA AAC Main Class
 * Primary API entry point for AAC sensor interaction
 */

import { WiaAacEventEmitter } from './EventEmitter';
import { SignalValidator, getSignalValidator } from './SignalValidator';
import { ISensorAdapter } from '../adapters/BaseAdapter';
import {
  WiaAacSignal,
  DeviceInfo,
  SensorType,
  EventType,
  EventHandler,
  EventHandlerOptions,
  EventDataMap,
  WiaAacError,
  ErrorCode,
  SensorConfig,
  WiaAacOptions,
  DEFAULT_OPTIONS,
  ConnectionState,
  SensorOptions
} from '../types';

// Logger interface
interface Logger {
  debug: (msg: string, ...args: unknown[]) => void;
  info: (msg: string, ...args: unknown[]) => void;
  warn: (msg: string, ...args: unknown[]) => void;
  error: (msg: string, ...args: unknown[]) => void;
}

// Adapter registry
const adapterRegistry: Map<SensorType, new () => ISensorAdapter> = new Map();

export class WiaAac {
  private options: Required<WiaAacOptions>;
  private emitter: WiaAacEventEmitter;
  private validator: SignalValidator;
  private adapter: ISensorAdapter | null = null;
  private config: SensorConfig | null = null;
  private state: ConnectionState = ConnectionState.DISCONNECTED;
  private signalBuffer: WiaAacSignal[] = [];
  private reconnectAttempts = 0;
  private reconnectTimer: NodeJS.Timeout | null = null;
  private logger: Logger;

  constructor(options: WiaAacOptions = {}) {
    this.options = { ...DEFAULT_OPTIONS, ...options };
    this.emitter = new WiaAacEventEmitter();
    this.validator = getSignalValidator();
    this.logger = this.createLogger();
  }

  // ============================================
  // Static Methods
  // ============================================

  /**
   * Register a custom adapter
   */
  static registerAdapter(type: SensorType, adapterClass: new () => ISensorAdapter): void {
    adapterRegistry.set(type, adapterClass);
  }

  /**
   * Get registered adapter class
   */
  static getAdapterClass(type: SensorType): (new () => ISensorAdapter) | undefined {
    return adapterRegistry.get(type);
  }

  // ============================================
  // Connection Management
  // ============================================

  /**
   * Connect to a sensor
   */
  async connect(config: SensorConfig): Promise<void> {
    if (this.state === ConnectionState.CONNECTED) {
      throw this.createError(ErrorCode.DEVICE_BUSY, 'Already connected to a device');
    }

    this.config = config;
    this.state = ConnectionState.CONNECTING;
    this.logger.info(`Connecting to ${config.type} sensor...`);

    try {
      // Get adapter (use pre-set adapter if available)
      if (!this.adapter) {
        const AdapterClass = adapterRegistry.get(config.type);
        if (!AdapterClass) {
          // Use mock adapter if no real adapter registered
          this.logger.warn(`No adapter registered for ${config.type}, using mock`);
          this.adapter = this.createMockAdapter(config.type);
        } else {
          this.adapter = new AdapterClass();
        }
      }

      // Connect
      await this.adapter.connect(config);

      // Setup signal handler
      this.adapter.onSignal((signal) => this.handleSignal(signal));

      // Apply options
      if (config.options) {
        this.adapter.configure(config.options);
      }

      this.state = ConnectionState.CONNECTED;
      this.reconnectAttempts = 0;
      this.logger.info(`Connected to ${config.device?.manufacturer || config.type}`);

      // Emit connected event
      const deviceInfo = this.adapter.deviceInfo || {
        manufacturer: config.device?.manufacturer || 'Unknown',
        model: config.device?.model || config.type
      };
      this.emitter.emit('connected', deviceInfo);

    } catch (error) {
      this.state = ConnectionState.ERROR;
      const wiaError = this.createError(
        ErrorCode.CONNECTION_FAILED,
        `Failed to connect: ${error instanceof Error ? error.message : 'Unknown error'}`
      );
      this.emitter.emit('error', wiaError);
      throw wiaError;
    }
  }

  /**
   * Disconnect from the sensor
   */
  async disconnect(): Promise<void> {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    if (this.adapter) {
      try {
        await this.adapter.disconnect();
      } catch (error) {
        this.logger.error('Error during disconnect:', error);
      }
      this.adapter = null;
    }

    this.state = ConnectionState.DISCONNECTED;
    this.config = null;
    this.signalBuffer = [];

    this.emitter.emit('disconnected', { reason: 'user' });
    this.logger.info('Disconnected');
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.state === ConnectionState.CONNECTED;
  }

  /**
   * Get connection state
   */
  getConnectionState(): ConnectionState {
    return this.state;
  }

  // ============================================
  // Device Discovery
  // ============================================

  /**
   * List available devices
   */
  async listDevices(): Promise<DeviceInfo[]> {
    // In a real implementation, this would scan for devices
    // For now, return empty list
    this.logger.debug('Listing devices...');
    return [];
  }

  /**
   * Get connected device info
   */
  getDeviceInfo(): DeviceInfo | null {
    return this.adapter?.deviceInfo || null;
  }

  // ============================================
  // Event Handling
  // ============================================

  /**
   * Subscribe to an event
   */
  on<T extends EventType>(
    event: T,
    handler: EventHandler<T>,
    options?: EventHandlerOptions<T>
  ): void {
    this.emitter.on(event, handler, options);
  }

  /**
   * Unsubscribe from an event
   */
  off<T extends EventType>(event: T, handler: EventHandler<T>): void {
    this.emitter.off(event, handler);
  }

  /**
   * Subscribe to an event once
   */
  once<T extends EventType>(event: T, handler: EventHandler<T>): void {
    this.emitter.once(event, handler);
  }

  // ============================================
  // Signal Management
  // ============================================

  /**
   * Get the last received signal
   */
  getLastSignal(): WiaAacSignal | null {
    return this.signalBuffer.length > 0
      ? this.signalBuffer[this.signalBuffer.length - 1]
      : null;
  }

  /**
   * Get signal buffer
   */
  getSignalBuffer(): WiaAacSignal[] {
    return [...this.signalBuffer];
  }

  // ============================================
  // Configuration
  // ============================================

  /**
   * Update sensor options
   */
  configure(options: Partial<SensorOptions>): void {
    if (this.adapter) {
      this.adapter.configure(options as SensorOptions);
    }
  }

  /**
   * Get current configuration
   */
  getConfig(): SensorConfig | null {
    return this.config;
  }

  // ============================================
  // Private Methods
  // ============================================

  private handleSignal(signal: WiaAacSignal): void {
    // Validate if enabled
    if (this.options.validateSignals) {
      if (!this.validator.isValidSignal(signal)) {
        this.logger.warn('Invalid signal received, skipping');
        return;
      }
    }

    // Add to buffer
    this.signalBuffer.push(signal);
    if (this.signalBuffer.length > this.options.signalBufferSize) {
      this.signalBuffer.shift();
    }

    // Emit signal event
    this.emitter.emit('signal', signal);

    // Check for selection events
    this.checkSelectionEvent(signal);
  }

  private checkSelectionEvent(signal: WiaAacSignal): void {
    // Eye tracker fixation -> selection
    if (signal.type === 'eye_tracker') {
      const fixation = signal.data.fixation;
      if (fixation?.active && fixation.duration_ms && fixation.duration_ms >= (this.config?.options?.dwellTime || 1000)) {
        this.emitter.emit('selection', {
          timestamp: signal.timestamp.unix_ms,
          targetId: fixation.target_id || 'unknown',
          targetType: 'area',
          selectionMethod: 'dwell',
          position: { x: signal.data.gaze.x, y: signal.data.gaze.y },
          confidence: signal.meta?.confidence || 0.8
        });
      }
    }

    // Switch press -> selection
    if (signal.type === 'switch' && signal.data.state === 'pressed') {
      this.emitter.emit('selection', {
        timestamp: signal.timestamp.unix_ms,
        targetId: `switch_${signal.data.switch_id}`,
        targetType: 'button',
        selectionMethod: 'switch',
        confidence: 1.0
      });
    }
  }

  private createError(code: ErrorCode, message: string): WiaAacError {
    return {
      code,
      message,
      timestamp: Date.now(),
      recoverable: code < 300
    };
  }

  private createLogger(): Logger {
    const level = this.options.logLevel;
    const levels = ['debug', 'info', 'warn', 'error', 'none'];
    const levelIndex = levels.indexOf(level);

    const noop = () => {};
    const log = (lvl: string) => (msg: string, ...args: unknown[]) => {
      console.log(`[WiaAac:${lvl}] ${msg}`, ...args);
    };

    return {
      debug: levelIndex <= 0 ? log('debug') : noop,
      info: levelIndex <= 1 ? log('info') : noop,
      warn: levelIndex <= 2 ? log('warn') : noop,
      error: levelIndex <= 3 ? log('error') : noop
    };
  }

  private createMockAdapter(type: SensorType): ISensorAdapter {
    // Import dynamically to avoid circular dependency
    const { MockAdapter } = require('../adapters/MockAdapter');
    return new MockAdapter({ type, simulateSignals: false });
  }

  /**
   * Use a specific adapter instance (for testing)
   */
  useAdapter(adapter: ISensorAdapter): void {
    this.adapter = adapter;
  }
}
